server/cmd/gram/deps.go

```diff
@@ -107,13 +106,42 @@ func newToolMetricsClient(ctx context.Context, logger *slog.Logger, c *cli.Conte
 
 	conn, err := clickhouse.Open(opts)
 	if err != nil {
-		logger.WarnContext(ctx, "error connecting to clickhouse; falling back to stub tool call metrics client")
-		return &tm.StubToolMetricsClient{}, func(context.Context) error { return nil }, nil
+		return nil, nilFunc, fmt.Errorf("failed to open clickhouse connection: %w", err)
+	}
+
+	// Retry ping with exponential backoff
```

Comment:
I haven't seen any lib we use for exponential backoffs (except the retryable http one, which isn't helpful here because we want to backoff around `conn.Ping`). If we do use something, let me know.

Reply:
Looking at it, I think it could be a bit of an overkill for this use case. I'll leave this as is for now as it's very explicit and a simple use case, we can revisit if this approach isn't working.

Thanks for the tip!